subsys/bluetooth/audio/aics.c

```diff
@@ -0,0 +1,824 @@
+		if (cp->gain_setting < inst->gain_settings.minimum ||
+		    cp->gain_setting > inst->gain_settings.maximum) {
+			return BT_GATT_ERR(BT_AICS_ERR_OUT_OF_RANGE);
+		}
+		if (BT_AICS_INPUT_MODE_SETTABLE(inst->state.gain_mode) &&
+		    inst->state.gain != cp->gain_setting) {
+			inst->state.gain = cp->gain_setting;
+			notify = true;
+		}
+		break;
+	case BT_AICS_OPCODE_UNMUTE:
+		BT_DBG("Unmute");
+		if (inst->state.mute == BT_AICS_STATE_MUTE_DISABLED) {
+			return BT_GATT_ERR(BT_AICS_ERR_MUTE_DISABLED);
+		}
+		if (inst->state.mute != BT_AICS_STATE_UNMUTED) {
+			inst->state.mute = BT_AICS_STATE_UNMUTED;
+			notify = true;
+		}
+		break;
+	case BT_AICS_OPCODE_MUTE:
+		BT_DBG("Mute");
+		if (inst->state.mute == BT_AICS_STATE_MUTE_DISABLED) {
+			return BT_GATT_ERR(BT_AICS_ERR_MUTE_DISABLED);
+		}
+		if (inst->state.mute != BT_AICS_STATE_MUTED) {
+			inst->state.mute = BT_AICS_STATE_MUTED;
+			notify = true;
+		}
+		break;
+	case BT_AICS_OPCODE_SET_MANUAL:
+		BT_DBG("Set manual mode");
+		if (BT_AICS_INPUT_MODE_IMMUTABLE(inst->state.gain_mode)) {
+			return BT_GATT_ERR(BT_AICS_ERR_GAIN_MODE_NOT_ALLOWED);
+		}
+		if (inst->state.gain_mode != BT_AICS_MODE_MANUAL) {
+			inst->state.gain_mode = BT_AICS_MODE_MANUAL;
+			notify = true;
+		}
+		break;
+	case BT_AICS_OPCODE_SET_AUTO:
+		BT_DBG("Set automatic mode");
+		if (BT_AICS_INPUT_MODE_IMMUTABLE(inst->state.gain_mode)) {
+			return BT_GATT_ERR(BT_AICS_ERR_GAIN_MODE_NOT_ALLOWED);
+		}
+		if (inst->state.gain_mode != BT_AICS_MODE_AUTO) {
+			inst->state.gain_mode = BT_AICS_MODE_AUTO;
+			notify = true;
+		}
+		break;
+	default:
+		return BT_GATT_ERR(BT_AICS_ERR_OP_NOT_SUPPORTED);
+	}
+
+	if (notify) {
+		inst->state.change_counter++;
+
+		BT_DBG("New state: gain %d, mute %u, gain_mode %u, counter %u",
+		       inst->state.gain, inst->state.mute,
+		       inst->state.gain_mode, inst->state.change_counter);
+
+		bt_gatt_notify_uuid(NULL, BT_UUID_AICS_STATE,
+				    inst->service_p->attrs, &inst->state,
+				    sizeof(inst->state));
+
+		if (inst->cb && inst->cb->state) {
+			inst->cb->state(NULL, (struct bt_aics *)inst, 0,
+					inst->state.gain, inst->state.mute,
+					inst->state.gain_mode);
+		} else {
+			BT_DBG("Callback not registered for instance %p", inst);
+		}
+	}
+
+	return len;
+}
+
+#if defined(CONFIG_BT_AICS)
+static void aics_description_cfg_changed(const struct bt_gatt_attr *attr,
+					 uint16_t value)
+{
+	BT_DBG("value 0x%04x", value);
+}
+#endif /* CONFIG_BT_AICS */
+
+static ssize_t write_description(struct bt_conn *conn,
+				 const struct bt_gatt_attr *attr,
+				 const void *buf, uint16_t len, uint16_t offset,
+				 uint8_t flags)
+{
+	struct bt_aics_server *inst = attr->user_data;
+
+	if (len >= sizeof(inst->description)) {
+		BT_DBG("Output desc was clipped from length %u to %zu",
+		       len, sizeof(inst->description) - 1);
+		/* We just clip the string value if it's too long */
+		len = (uint16_t)sizeof(inst->description) - 1;
+	}
```

Comment:
Here, as everywhere else (also in other services), we have the issue of possibly truncating the string in the middle of a more-than-one-octet letter.
I have no good solution to this, and do not propose to change this, so this comment is just a reminder to ourselves about this issue.

Reply:
I guess what we need is a `truncate_utf8_str(const char *utf8_in, size_t size, char *utf8_out)` :) Might need to be `uint8_t *` pointers instead, as appending the NULL terminator might not be easy in all cases.

Reply:
Pretty much something like https://stackoverflow.com/questions/35328529/stdstring-optimal-way-to-truncate-utf-8-at-safe-place

i.e. iterate the byte array to find a byte that's the beginning of a utf8 char and cut it there

Reply:
https://github.com/zephyrproject-rtos/zephyr/issues/34159